library/init/category/state.lean

```diff
@@ -1,124 +1,147 @@
+  ⟨λ s, x₁.run s <|> x₂.run s⟩
+
+  protected def failure [alternative m] {α : Type u} : state_t σ m α :=
+  ⟨λ s, failure⟩
+
+  instance [alternative m] : alternative (state_t σ m) :=
+  { failure := @state_t.failure _ _ _ _,
+    orelse  := @state_t.orelse _ _ _ _ }
+
+  @[inline] protected def get : state_t σ m σ :=
+  ⟨λ s, pure (s, s)⟩
+
+  @[inline] protected def put : σ → state_t σ m punit :=
+  λ s', ⟨λ s, pure (punit.star, s')⟩
+
+  @[inline] protected def modify (f : σ → σ) : state_t σ m punit :=
+  ⟨λ s, pure (punit.star, f s)⟩
+
+  @[inline] protected def lift {α : Type u} (t : m α) : state_t σ m α :=
+  ⟨λ s, do a ← t, pure (a, s)⟩
 
-  def state_t_return (a : α) : state_t σ m α :=
-  λ s, show m (α × σ), from
-    return (a, s)
+  instance : has_monad_lift m (state_t σ m) :=
+  ⟨@state_t.lift σ m _⟩
 
-  def state_t_bind (act₁ : state_t σ m α) (act₂ : α → state_t σ m β) : state_t σ m β :=
-  λ s, show m (β × σ), from
-     do (a, new_s) ← act₁ s,
-        act₂ a new_s
+  @[inline] protected def monad_map {σ m m'} [monad m] [monad m'] {α} (f : Π {α}, m α → m' α) :
+    state_t σ m α → state_t σ m' α :=
+  λ x, ⟨λ st, f (x.run st)⟩
+
+  instance (σ m m') [monad m] [monad m'] : monad_functor m m' (state_t σ m) (state_t σ m') :=
+  ⟨@state_t.monad_map σ m m' _ _⟩
+
+  -- TODO(Sebastian): uses lenses as in https://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-Zoom.html#t:Zoom ?
+  protected def zoom {σ σ' α : Type u} {m : Type u → Type v} [monad m] (get : σ → σ')
+    (set : σ' → σ → σ) (x : state_t σ' m α) : state_t σ m α :=
+  ⟨λ st, do (a, st') ← x.run (get st),
+            pure (a, set st' st)⟩
+
+  instance (ε) [monad_except ε m] : monad_except ε (state_t σ m) :=
+  { throw := λ α, state_t.lift ∘ throw,
+    catch := λ α x c, ⟨λ s, catch (x.run s) (λ e, state_t.run (c e) s)⟩ }
 end
+end state_t
 
-instance (σ : Type u) (m : Type u → Type v) [monad m] : monad (state_t σ m) :=
-{pure := @state_t_return σ m _, bind := @state_t_bind σ m _,
- id_map := begin
-   intros, funext,
-   simp [state_t_bind, state_t_return, function.comp, return],
-   have h : state_t_bind._match_1 (λ (x : α) (s : σ), @pure m _ _ (x, s)) = pure,
-   { funext s, cases s, apply rfl },
-   { simp [h, monad.bind_pure] },
- end,
- pure_bind := begin
-   intros, funext,
-   simp [state_t_bind, state_t_return, monad.pure_bind]
- end,
- bind_assoc := begin
-   intros, funext,
-   simp [state_t_bind, state_t_return, monad.bind_assoc],
-   apply congr_arg, funext r,
-   cases r, refl
- end}
+/-- A specialization of `monad_lift` to lifting `state_t` that allows `σ` to be inferred.
+
+    This class is roughly equivalent to `MonadState` from https://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-State-Class.html,
+    with the important distinction that it is automatically derived via the generic
+    `has_monad_lift` class. -/
+class monad_state_lift (σ : out_param (Type u)) (m : out_param (Type u → Type v)) (n : Type u → Type w) :=
```

Comment:
I think we should say this is an auxiliary class, and users should not define instances for it.
My understanding is that this is a workaround to address a limitation of the type class resolution procedure.

Reply:
I'm not sure I would call it a workaround; I can't think of a better solution in a cleverer typeclass system. I'm deriving instances of it for encapsulation https://github.com/Kha/lean/blob/tactic_trans/library/init/meta/interaction_monad.lean#L52